### PR TITLE
[Fabric] 2.x chaincode lifecycle update

### PR DESCRIPTION
**Pull request by himanshulalarya**
Signed-off-by: Lal <himanshu.a.lal@accenture.com>

**Changelog**
- Add Chaincode commit
- Updated Fabric 2.x lifecycle update

 

**Reviewed by**
@jagpreetsinghsasan 

 

**Linked issue**
#980

### CODE REVIEW

File: platforms/corda-ent/configuration/roles/setup/nmap/tasks/main.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
+##############################################################################################
+# This role creates the value file for nmap
```
Comment (sownak) [inline platforms/corda-ent/configuration/roles/setup/nmap/tasks/main.yml:2]: merging needs to be fixed

**Review by sownak — CHANGES_REQUESTED** [review]

File: platforms/hyperledger-fabric/charts/commit_chaincode/templates/_helpers.tpl [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{{- define "labels.custom" }}
+  {{ range $key, $val := $.Values.metadata.labels }}
+  {{ $key }}: {{ $val }}
+  {{ end }}
+{{- end }}
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/charts/commit_chaincode/templates/_helpers.tpl:5]: Add newline

File: platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml [hunk 3]
```diff
@@ -50,6 +50,8 @@ spec:
           value: "{{ $.Values.vault.orderersecretprefix }}"
         - name: MOUNT_PATH
           value: /secret
+        - name: NETWORK_VERSION
+          value: "{{ $.Values.metadata.network.version}}"
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml:54]: Make this variable from {{ VAR}} to {{ VAR }}

File: platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml [hunk 4]
```diff
@@ -152,8 +154,23 @@ spec:
...
+            peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/chaincode/${CHAINCODE_NAME}/${CHAINCODE_MAINDIR}
+            echo "1.4 cmd executed"
+          {{ else }}
+            echo "Inside 2.x chaincode"
+            echo ${NETWORK_VERSION} 
+            ## packaging chaincode
+            peer lifecycle chaincode package ${CHAINCODE_NAME}.tar.gz --path github.com/chaincode/${CHAINCODE_NAME}/${CHAINCODE_MAINDIR} --lang ${CC_RUNTIME_LANGUAGE} --label ${CHAINCODE_NAME}_${CHAINCODE_VERSION}
+            ## Installing Chaincode
+            peer lifecycle chaincode install ${CHAINCODE_NAME}.tar.gz
+            echo "2.x cmd executed"
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml:171]: change this to echo "Chaincode installed for Fabric 2.x"

File: platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml [hunk 5]
```diff
@@ -152,8 +154,23 @@ spec:
...
-          peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/chaincode/${CHAINCODE_NAME}/${CHAINCODE_MAINDIR}
+
+          {{ if eq  $.Values.metadata.network.version "1.4.4" }}
+            echo "Inside 1.4 chaincode"
+            echo ${NETWORK_VERSION} 
+            ## Installing chaincode on a peer
+            peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/chaincode/${CHAINCODE_NAME}/${CHAINCODE_MAINDIR}
+            echo "1.4 cmd executed"
+          {{ else }}
+            echo "Inside 2.x chaincode"
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml:165]: Change this echo to something meaningful

File: platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml [hunk 6]
```diff
@@ -152,8 +154,23 @@ spec:
           # Get dependencies
           mkdir -p $GOPATH/bin && curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
           cd $GOPATH/src/github.com/chaincode && dep ensure
-          ## Installing chaincode on a peer
-          peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/chaincode/${CHAINCODE_NAME}/${CHAINCODE_MAINDIR}
+
+          {{ if eq  $.Values.metadata.network.version "1.4.4" }}
+            echo "Inside 1.4 chaincode"
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml:159]: Change this echo to something meaningful

File: platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml [hunk 7]
```diff
@@ -152,8 +154,23 @@ spec:
...
+            echo ${NETWORK_VERSION}
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml:160]: Remove this

File: platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml [hunk 8]
```diff
@@ -152,8 +154,23 @@ spec:
...
           cd $GOPATH/src/github.com/chaincode && dep ensure
-          ## Installing chaincode on a peer
-          peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/chaincode/${CHAINCODE_NAME}/${CHAINCODE_MAINDIR}
+
+          {{ if eq  $.Values.metadata.network.version "1.4.4" }}
+            echo "Inside 1.4 chaincode"
+            echo ${NETWORK_VERSION} 
+            ## Installing chaincode on a peer
+            peer chaincode install -n ${CHAINCODE_NAME} -v ${CHAINCODE_VERSION} -p github.com/chaincode/${CHAINCODE_NAME}/${CHAINCODE_MAINDIR}
+            echo "1.4 cmd executed"
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/charts/install_chaincode/templates/install_chaincode.yaml:163]: Change this to echo "Chaincode install for Fabric v.1.4.4"

File: platforms/hyperledger-fabric/configuration/roles/create/chaincode/commit/tasks/main.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+# Create value file for chaincode commit
+- name: "Create value file for chaincode commit"
+  include_tasks: valuefile.yaml
+  vars:
+    channelcreator_query: "participants[?type=='creator']"
+    org_query: "organizations[?name=='{{participant.name}}']"
+    org: "{{ network | json_query(org_query) | first }}"
+  loop: "{{ item | json_query(channelcreator_query) }}"
+  loop_control:
+    loop_var: participant
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/configuration/roles/create/chaincode/commit/tasks/main.yaml:16]: Add newline

File: platforms/hyperledger-fabric/configuration/roles/create/orderers/vars/main.yaml [hunk 10]
```diff
@@ -2,7 +2,9 @@ kafka_image_version:
   1.4.0: 0.4.14
   1.4.4: 0.4.18
   2.0.0: 0.4.18
+  2.2.0: 0.4.18
 zookeeper_image_version:
   1.4.0: 0.4.14
   1.4.4: 0.4.18
   2.0.0: 0.4.18
+  2.2.0: 0.4.18
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/configuration/roles/create/orderers/vars/main.yaml:10]: Add newline

File: platforms/hyperledger-fabric/configuration/roles/helm_component/templates/commit_chaincode_job.tpl [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    orderer:
+      address: {{ participant.ordererAddress }}
+    chaincode:
+      builder: hyperledger/fabric-ccenv:{{ network.version }}
+      name: {{ component_chaincode.name | lower | e }}
+      version: {{ component_chaincode.version }}
+      commitarguments: {{ component_chaincode.arguments | quote}}
+      endorsementpolicies:  {{ component_chaincode.endorsements | quote}}
+    channel:
+      name: {{ item.channel_name | lower }}
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/configuration/roles/helm_component/templates/commit_chaincode_job.tpl:44]: Add newline

File: platforms/hyperledger-fabric/configuration/samples/network-fabric-add-organization.yaml [hunk 12]
```diff
@@ -415,6 +421,7 @@ network:
             name: "chaincode_name" #This has to be replaced with the name of the chaincode
             version: "chaincode_version" #This has to be replaced with the version of the chaincode
             maindirectory: "chaincode_main"  #The main directory where chaincode is needed to be placed
+            lang: "golang"  # The language in which the chaincode is written ( golang/ java/ nodejs)
```
Comment (jagpreetsinghsasan) [inline platforms/hyperledger-fabric/configuration/samples/network-fabric-add-organization.yaml:424]: Dont put nodejs as an option until we have it

**Review by jagpreetsinghsasan — CHANGES_REQUESTED** [review]

**Comment by jagpreetsinghsasan** [comment]
Make sure to always squash and merge all commits into 1

**Review by jagpreetsinghsasan — APPROVED** [review]
Just make sure to have end-to-end testing before this gets merged. The code seems fine now.

File: platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml [hunk 13]
```diff
@@ -127,7 +127,12 @@ spec:
           export ARGS='{"Args":'"'["${INSTANTIATION_ARGUMENTS}"]'"'}'
           export qARGS="'${ARGS}'"
           touch ./invokeChaincode.sh
-          echo "peer chaincode invoke -o ${ORDERER_URL} --tls ${CORE_PEER_TLS_ENABLED} --cafile ${ORDERER_CA} -C ${CHANNEL_NAME} -n ${CHAINCODE_NAME} -c $qARGS" >> ./invokeChaincode.sh
+          {{ if eq  $.Values.metadata.network.version "1.4.4" }}
```
Comment (sownak) [inline platforms/hyperledger-fabric/charts/invoke_chaincode/templates/invoke_chaincode.yaml:130]: why hardcoded to 1.4.4?

File: platforms/hyperledger-fabric/configuration/roles/delete/flux_releases/meta/main.yaml [hunk 14]
```diff
@@ -3,6 +3,7 @@ dependencies:
   - role: "{{playbook_dir}}/../../shared/configuration/roles/setup/helm"
     vars:
       helm:
+        version: "v2.14.1"
```
Comment (sownak) [inline platforms/hyperledger-fabric/configuration/roles/delete/flux_releases/meta/main.yaml:6]: need to be removed

File: platforms/hyperledger-fabric/configuration/roles/helm_component/templates/install_chaincode_job.tpl [hunk 15]
```diff
@@ -39,6 +41,7 @@ spec:
       builder: hyperledger/fabric-ccenv:{{ network.version }}
       name: {{ component_chaincode.name | lower | e }}
       version: {{ component_chaincode.version }}
+      lang: {{ component_chaincode.lang }}
```
Comment (sownak) [inline platforms/hyperledger-fabric/configuration/roles/helm_component/templates/install_chaincode_job.tpl:44]: if this fails when lang does not exist in network.yaml, then that needs to be corrected.

**Review by sownak — CHANGES_REQUESTED** [review]

**Comment by sownak** [comment]
@jagpreetsinghsasan Please resolve conversation after checking the updates

**Review by jagpreetsinghsasan — APPROVED** [review]

**Review by sownak — APPROVED** [review]

**Comment by sownak** [comment]
Approved, but will be merged after current release.

**Pull request merged**